runtime/src/iree/io/formats/irpa/irpa_builder.h

```diff
@@ -0,0 +1,123 @@
+//
+// Example:
+//  iree_io_parameter_archive_builder_t builder;
+//  iree_io_parameter_archive_builder_initialize(host_allocator, &builder);
+//  iree_io_parameter_archive_builder_add_data_entry(&builder, ...);
+//  iree_io_parameter_archive_builder_add_data_entry(&builder, ...);
+//  iree_io_parameter_archive_builder_add_data_entry(&builder, ...);
+//  total_size = iree_io_parameter_archive_builder_total_size(&builder);
+//  << create file of total_size, map into memory >>
+//  iree_io_parameter_archive_builder_write(&builder, file, &target_index);
+//  << file now contains the full archive header >>
+//  << target_index now references the ranges in the file >>
+//  << write parameter contents, or don't if leaving uninitialized >>
+//  iree_io_parameter_archive_builder_deinitialize(&builder);
+typedef struct iree_io_parameter_archive_builder_t {
+  iree_allocator_t host_allocator;
+  iree_io_parameter_index_t* index;
+  iree_io_physical_size_t file_alignment;
+  iree_io_physical_size_t entry_segment_size;
+  iree_io_physical_size_t metadata_segment_size;
+  iree_io_physical_size_t storage_segment_size;
+  iree_io_physical_size_t storage_alignment;
+} iree_io_parameter_archive_builder_t;
+
+// Initializes a new parameter builder in |out_builder| for use.
+// iree_io_parameter_archive_builder_deinitialize must be called to drop any
+// resources retained while building.
+IREE_API_EXPORT iree_status_t iree_io_parameter_archive_builder_initialize(
+    iree_allocator_t host_allocator,
+    iree_io_parameter_archive_builder_t* out_builder);
+
+// Deinitializes |builder| and drops all resources.
+IREE_API_EXPORT void iree_io_parameter_archive_builder_deinitialize(
+    iree_io_parameter_archive_builder_t* builder);
+
+// Returns the total file size required to store the parameter archive header
+// and contents of all added parameters. Adding new parameters will invalidate
+// this value.
+IREE_API_EXPORT iree_io_physical_size_t
+iree_io_parameter_archive_builder_total_size(
+    const iree_io_parameter_archive_builder_t* builder);
+
+// Writes the parameter archive to the given |file_handle|. The file must have
+// at least enough storage to fit iree_io_parameter_archive_builder_total_size.
+// The archive will be written starting at the given |file_offset|.
+// If an optional |target_index| is provided entries for all parameters will be
+// appended to the index referencing the given |file_handle|.
+IREE_API_EXPORT iree_status_t iree_io_parameter_archive_builder_write(
+    const iree_io_parameter_archive_builder_t* builder,
+    iree_io_file_handle_t* file_handle, iree_io_physical_offset_t file_offset,
+    iree_io_stream_t* stream, iree_io_parameter_index_t* target_index);
+
+// Adds a new splat entry to |builder|.
+// Splat entries have no physical storage and exist only in the header.
+// |pattern| and |metadata| (if provided) are copied prior to returning.
+IREE_API_EXPORT iree_status_t iree_io_parameter_archive_builder_add_splat_entry(
```

Comment:
Maybe a note on maximum splat pattern size? I've mis-assumed this before and it cost me time...